CommitPoolContract/contracts/SinglePlayerCommit.sol

```diff
@@ -406,31 +234,205 @@ contract SinglePlayerCommit is ChainlinkClient, Ownable {
+    }
+
+    // @notice Enables processing of open commitments after endDate that have not been processed by creator
+    // @param commitmentOwner address of the creator of the committer to process
+    // @dev Process commitment by lookup based on address, checking metrics, state and updating balances
+    function _processCommitment(address commitmentOwner) internal {
+        console.log("Processing commitment");
+        Commitment storage commitment = commitments[commitmentOwner];
+
+        require(commitment.endTime < block.timestamp, "SPC::processCommitment - commitment is still ongoing");
+        require(commitment.endTime < commitment.lastActivityUpdate, "SPC::processCommitment - update activity");
+        
+        commitment.met = commitment.reportedValue >= commitment.goalValue;
+
+        emit CommitmentEnded(commitmentOwner, commitment.met, commitment.stake);
+    }
+
+    // @notice Enables control of processing own commitment. For instance when completed.
+    // @dev Process commitment by lookup msg.sender, checking metrics, state and updating balances
+    function _processCommitmentUser() internal activeCommitment {
+        console.log("Processing commitment");
+        Commitment storage commitment = commitments[msg.sender];
+
+        commitment.met = commitment.reportedValue >= commitment.goalValue;
+
+        emit CommitmentEnded(msg.sender, commitment.met, commitment.stake);
+    }
+
+    // @notice Internal function for evaluating commitment and slashing funds if needed
+    // @dev Receive call with commitment object from storage
+    function _settleCommitment(address commitmentOwner) internal activeCommitment {
+        console.log("Settling commitment");
+        Commitment storage commitment = commitments[commitmentOwner];
+        commitment.exists = false;        
+        commitment.met ? _withdraw(commitment.stake, commitmentOwner) : _slashFunds(commitment.stake, commitmentOwner);
+    }
+
+    // @notice Internal function to slash funds from user
+    // @param amount Amount of <token> to slash
+    // @param committer Address of committer
+    // @dev Substract amount from committer balance and add to slashedBalance
+    function _slashFunds(uint256 amount, address committer) internal noCommitment {
+        console.log("Slashing funds commitment");
+
+        require(committerBalances[committer] >= amount, "SPC::_slashFunds - funds not available");
+
+        _changeCommitterBalance(committer, amount, false);
+        slashedBalance = slashedBalance.add(amount);
+        emit FundsSlashed(committer, amount, slashedBalance);
+    }
+
+    // @notice Public function to withdraw unstaked balance of user
+    // @param amount Amount of <token> to withdraw
+    /// @dev Check balances and active stake, withdraw from balances, emit event
+    function _withdraw(uint256 amount, address committer) internal {
```

Comment:
Since this function is only called by `_settleCommitment`, I wonder if it makes sense to get rid of the require checks and other preliminary logic. Since we're *only* withdrawing the returned stake from a specific commitment, technically we don't need to check the global balance for the committer, which means we can also get rid of the available calculations.

I'm thinking we can do something like this...

```
function _withdraw(uint256 amount, address committer) internal {
    _changeCommitterBalance(committer, amount, false);
    
    require(token.transfer(committer, amount), "SPC::withdraw - token transfer failed");

    emit Withdrawal(committer, amount, committerBalances[committer]);
} 
```

The other option here is to forgo using a separate function altogether and embed those three lines into `_settleCommitment` directly. Once we enable multiple commitments per user, we would bring back the available balance checks.

Or, since on Matic gas is cheap af maybe we don't need to worry about this at all until we establish a more permament architecture...thoughts?

Reply:
The main thought I had here now is that, for the user, it is not possible to withdraw their funds now.. Unless we say you deposit what you stake and you withdraw what you stake. I think that could fit/work for the SinglePlayer mode, but if we're going to think about a generic contract with another contract that is the SPC implementation, that should change.

My general thought is that calling internal functions is free, and when omitting the events the functions are quite clear. But also a little to much for the scope that we have now :) I'll do some updates this afternoon. If the function stays easy to read I'm in also in favor of mergin settle, withdraw & slash